src/hotspot/share/opto/predicates.cpp

```diff
@@ -147,3 +149,126 @@ Node* PredicateBlock::skip_regular_predicates(Node* regular_predicate_proj, Deop
+// to and including the OpaqueLoop* nodes). The cloned nodes are rewired to reflect the same graph structure as found for
+// this Template Assertion Predicate Expression. The cloned nodes get 'new_ctrl' as ctrl. There is no other update done
+// for the cloned nodes. Return the newly cloned Opaque4Node.
+Opaque4Node* TemplateAssertionPredicateExpression::clone(Node* new_ctrl, PhaseIdealLoop* phase) {
+  CloneStrategy clone_init_and_stride_strategy(phase, new_ctrl);
+  return clone(clone_init_and_stride_strategy, new_ctrl, phase);
+}
+
+// Class to collect data nodes from a source to target nodes by following the inputs of the source node recursively.
+// The class takes a node filter to decide which input nodes to follow and a target node predicate to start backtracking
+// from. All nodes found on all paths from source->target(s) are returned in a Unique_Node_List (without duplicates).
+class DataNodesOnPathsToTargets : public StackObj {
+  typedef bool (*NodeCheck)(const Node*);
+
+  // Node filter function to decide if we should process a node or not while searching for targets.
+  NodeCheck _node_filter;
+  // Function to decide if a node is a target node (i.e. where we should start backtracking). This check should also
+  // trivially pass the _node_filter.
+  NodeCheck _is_target_node;
+  Unique_Node_List _collected_nodes; // The resulting node collection of all nodes on paths from source->target(s).
+  // List to track all nodes visited on the search for target nodes starting at a start node. These nodes are then used
+  // in backtracking to find the nodes actually being on a start->target(s) path. This list also serves as visited set
+  // to avoid double visits of a node which could happen with diamonds shapes.
+  Unique_Node_List _nodes_to_visit;
+
+ public:
+  DataNodesOnPathsToTargets(NodeCheck node_filter, NodeCheck is_target_node)
+      : _node_filter(node_filter),
+        _is_target_node(is_target_node) {}
+  NONCOPYABLE(DataNodesOnPathsToTargets);
+
+  // Collect all input nodes from 'start_node'->target(s) by applying the node filter to discover new input nodes and
+  // the target node predicate to stop discovering more inputs and start backtracking. The implementation is done
+  // with two BFS traversal: One to collect the target nodes (if any) and one to backtrack from the target nodes to
+  // find all other nodes on the start->target(s) paths.
+  const Unique_Node_List& collect(Node* start_node) {
+    assert(_collected_nodes.size() == 0 && _nodes_to_visit.size() == 0, "should not call this method twice in a row");
+    assert(!_is_target_node(start_node), "no trivial paths where start node is also a target node");
+
+    collect_target_nodes(start_node);
+    if (_collected_nodes.size() != 0) {
+      backtrack_from_target_nodes();
+      assert(_collected_nodes.member(start_node), "must find start node again when backtracking");
+    }
```

Comment:
What scenario is there where we collect no nodes? Probably there is some, where we don't have to clone anything.. but it's a bit strange. Would be nice to have a quick comment here about that.

For simplicity, you could always call backtrack (it just does nothing anyway).
Then you can just make the assert a bit smarter:
`assert(_collected_nodes.size() == 0 || _collected_nodes.member(start_node), "must find start node again when backtracking");`